---
title: "Introduction"
author: "Yanxiang Deng"
date: "01/29/2019"
output: html_document
---

# seq-Explorer

seq-Explorer is an integrated and interactive software platform for single-cell RNA-seq data analysis. It integrates **(1) quality control for high throughput sequence data (FastQC)**, **(2) digital gene expression matrix generation (Drop-seq)**,  **(3) mixed-species data processing (Drop-seq)**, and **(4) single cell RNA-seq data analysis (Seurat)**. In addition, the software is built on [R shiny](https://shiny.rstudio.com/), in which users can take an interactive approach to explore and visualize the data. The platform consists of the four modules:


### 1. FastQC

[FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) provide a tool for quality control checks on raw sequence data generated by high-throughput sequencer. seq-Explorer run FastQC only on **HPC**, but can visualize results in any system.

### 2. DGE

This module uses [Drop-seq pipeline](http://mccarrolllab.org/dropseq/) to generate digital gene expression matrix from raw sequence data. More details can be found in [this protocol](https://github.com/broadinstitute/Drop-seq/releases). seq-Explorer run DGE only on **HPC**, but can visualize results in any system. seq-Explorer uses [Drop-seq_tools-2.1.0](https://github.com/broadinstitute/Drop-seq/releases).

### 3. Species mix

This module uses [Drop-seq pipeline](http://mccarrolllab.org/dropseq/) to split Human and Mouse data from mixed species. It can process three file types: (1) BAM file, (2) summary file, amd (3) DGE file. seq-Explorer process BAM file only on **HPC**, but can process other two file types in any system. Two separated dge files for each species will be generated from BAM file or DGE file. More details can be found in [this protocol](https://github.com/broadinstitute/Drop-seq/releases). 

### 4. Seurat

[Seurat](https://satijalab.org/seurat/) is designed for QC, analysis, and exploration of single cell RNA-seq data. seq-Explorer has two submodules: (1) [normal workflow](https://satijalab.org/seurat/pbmc3k_tutorial.html) and (2) [integrated analyses](https://satijalab.org/seurat/immune_alignment.html). Two file types: (1) DGE file, (2) 10x Genomics file can be processed. In addition, data from the [online tutorial](https://satijalab.org/seurat/get_started.html) can be used for practice. seq-Explorer uses [Seurat 2.3.4](https://satijalab.org/seurat/install.html).

### 5. References
[1] https://shiny.rstudio.com/ <br/>
[2] https://www.bioinformatics.babraham.ac.uk/projects/fastqc/ <br/>
[3] [Macosko et al, **Cell** 2015](https://www.cell.com/abstract/S0092-8674%2915%2900549-8) <br/>
[4] [Butler et al., **Nature Biotechnology** 2018](https://www.nature.com/articles/nbt.4096) <br/>
[5] https://github.com/nasqar/SeuratWizard

### 6. License
seq-Explorer is developed and maintained by Yanxiang Deng@[Rong Fan Lab](https://www.eng.yale.edu/fanlab/Rong_Fan_Group/Welcome.html), BME, Yale University. The software is licensed under the GPL-3.0.

### 7. Contact
seq-Explorer is hosted at [Github](https://github.com/dyxmvp/seqExplorer). Please e-mail [yanxiang.deng@yale.edu](mailto:yanxiang.deng@yale.edu) or submit pull requests from the [Github](https://github.com/dyxmvp/seqExplorer) with any questions about the software.
